maven-extension/src/main/java/io/opentelemetry/maven/OtelExecutionListener.java

```diff
@@ -0,0 +1,248 @@
+ * execution of the last project ({@link #projectSucceeded(ExecutionEvent)} and {@link
+ * #projectFailed(ExecutionEvent)}) rather than on the end of the Maven session {@link
+ * #sessionEnded(ExecutionEvent)} because OpenTelemetry and GRPC classes are unloaded by the Maven
+ * classloader before {@link #sessionEnded(ExecutionEvent)} causing {@link NoClassDefFoundError}
+ * messages in the logs.
+ */
+@Component(role = ExecutionListener.class, hint = "otel-execution-listener")
+public class OtelExecutionListener extends AbstractExecutionListener {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  @Requirement private SpanRegistry spanRegistry;
+
+  @Requirement private OpenTelemetrySdkService openTelemetrySdkService;
+
+  /**
+   * Register in given {@link OtelExecutionListener} to the lifecycle of the given {@link
+   * MavenSession}
+   *
+   * @see org.apache.maven.execution.MavenExecutionRequest#setExecutionListener(ExecutionListener)
+   */
+  public static void registerOtelExecutionListener(
+      @Nonnull MavenSession session, @Nonnull OtelExecutionListener otelExecutionListener) {
+    @Nullable
+    ExecutionListener initialExecutionListener = session.getRequest().getExecutionListener();
+    if (initialExecutionListener instanceof ChainedExecutionListener) {
+      // already initialized
+      LoggerFactory.getLogger(OtelExecutionListener.class)
+          .debug(
+              "OpenTelemetry: OpenTelemetry extension already registered as execution listener, skip.");
+    } else {
+      session
+          .getRequest()
+          .setExecutionListener(
+              new ChainedExecutionListener(otelExecutionListener, initialExecutionListener));
+      LoggerFactory.getLogger(OtelExecutionListener.class)
+          .debug(
+              "OpenTelemetry: OpenTelemetry extension registered as execution listener. InitialExecutionListener: "
+                  + initialExecutionListener);
+    }
+  }
+
+  @Override
+  public void sessionStarted(ExecutionEvent executionEvent) {
+    MavenProject project = executionEvent.getSession().getTopLevelProject();
+    TextMapGetter<Map<String, String>> getter =
+        new TextMapGetter<Map<String, String>>() {
+          @Override
+          public Iterable<String> keys(Map<String, String> environmentVariables) {
+            return environmentVariables.keySet();
+          }
+
+          @Nullable
+          @Override
+          public String get(@Nullable Map<String, String> environmentVariables, String key) {
+            return environmentVariables == null
+                ? null
+                : environmentVariables.get(key.toUpperCase(Locale.ROOT));
+          }
+        };
+    io.opentelemetry.context.Context context =
+        W3CTraceContextPropagator.getInstance()
+            .extract(io.opentelemetry.context.Context.current(), System.getenv(), getter);
+    try (Scope scope = context.makeCurrent()) {
```

Comment:
A scope is a threadlocal that is generally meant to be accessed in some downstream code (e.g. in a servlet filter type of api, `.next()` would be called within the scope). There's no reason to make the scope current with this sort of pattern since there is no downstream code being called.

It seems that instead of threadlocal context propagation, we get around it by always parenting projects to the root span, and always parenting mojos to the project. This is OK and we could consider improving in the future.

The main issue is that other mojos won't be able to add their own data to the trace if they ever get instrumented. If Maven's executor _always_ processes a project in a single thread (e.g., it gets multi-threading by dispatching projects across threads, and it's guaranteed to be a contiguous start - finish of a single project on a single thread), then in `projectStarted` we would want to `makeCurrent`() without using try / resources (highly unsafe which is why we need to know the threading model correctly) and close the scope in the projectSucceeded / Failed.

Unless you're sure of the threading model, let's definitely not try that in this PR

Reply:
Comment added, I use `rootSpan.makeCurrent()` so it is defined as the parent span of the `projectSpan` that is created just after.

Reply:
Ah you should just use `SpanBuilder.setParent`, no need to make current just to set the parent.